### PR TITLE
feat: get staffs

### PR DESCRIPTION
**Pull request by nguyennh4522**
#### What's this PR do?

- [x] feat: get staffs

#### cURL

```curl
curl --location --request GET 'localhost:8080/api/v1/projects/8dc3be2e-19a4-4942-8a79-56db391a0b15/staffs'
```

#### Screenshots (if appropriate)

image

### CODE REVIEW

File: pkg/handler/project/project.go [hunk 1]
```diff
@@ -132,3 +135,80 @@ func (h *handler) UpdateProjectStatus(c *gin.Context) {
...
 }
+
+// GetStaffs godoc
+// @Summary Get list staffs of project
+// @Description Get list staffs of project
+// @Tags Project
+// @Accept  json
+// @Produce  json
+// @Param Authorization header string true "jwt token"
+// @Param project_id	path 	string	false  "Project status"
```
Comment (datphamcode295) [inline pkg/handler/project/project.go:146]: "Project status" -> "Project ID"

File: pkg/handler/project/project.go [hunk 2]
```diff
@@ -132,3 +135,80 @@ func (h *handler) UpdateProjectStatus(c *gin.Context) {
...
+// @Param status   query  string	false  "Status"
+// @Param page   query  string	false  "Page"
+// @Param size   query  string	false  "Size"
+// @Param sort   query  string	false  "Sort"
+// @Success 200 {object} view.ProjectMemberListResponse
+// @Failure 400 {object} view.ErrorResponse
```
Comment (datphamcode295) [inline pkg/handler/project/project.go:152]: add error 404 NOT FOUND

File: pkg/handler/project/project.go (hunk 1 above)
Comment (datphamcode295) [inline pkg/handler/project/project.go:146]:
```suggestion
// @Param id	path 	string	false  "Project ID"
```

File: pkg/handler/project/project.go (hunk 2 above)
Comment (nguyennh4522) [reply pkg/handler/project/project.go:151]: updated

File: pkg/store/project_slot/filter.go [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package projectslot
```
Comment (namnhce) [inline pkg/store/project_slot/filter.go:1]: folder name `pkg/store/project_slot` ---> `pkg/store/projectslot`

File: pkg/view/project.go [hunk 4]
```diff
@@ -32,13 +32,19 @@ type UpdatedProject struct {
...
+	Avatar         string     `json:"avatar"`
+	Position       string     `json:"position"`
+	Seniority      string     `json:"seniority"`
+	Status         string     `json:"status"`
+	IsLead         bool       `json:"isLead"`
+	DeploymentType string     `json:"deploymentType"`
+	JoinedDate     *time.Time `json:"joinedDate"`
+	LeftDate       *time.Time `json:"leftDate"`
+
+	Roles []RoleData `json:"roles"`
```
Comment (namnhce) [inline pkg/view/project.go:47]: Do we need Roles Here?

**Review by namnhce — CHANGES_REQUESTED** [review]

File: pkg/store/project_head/filter.go [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package projecthead
```
[the remark this answers is not in the document]
Comment (nguyennh4522) [reply pkg/store/project_head/filter.go:1]: updated

File: migrations/schemas/20221108214100-init_projects_schema.sql [hunk 6]
```diff
@@ -49,7 +49,7 @@ CREATE TABLE IF NOT EXISTS project_slots (
     created_at       TIMESTAMP(6)     DEFAULT (NOW()),
     updated_at       TIMESTAMP(6)     DEFAULT (NOW()),
     project_id       uuid NOT NULL,
-    seniority_id     uuid NOT NULL,
```
[the remark this answers is not in the document]
Comment (nguyennh4522) [reply migrations/schemas/20221108214100-init_projects_schema.sql]: updated

File: pkg/store/project_head/project.go [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+func New(db *gorm.DB) IStore {
+	return &store{
+		db: db,
+	}
+}
+
+// GetByProjectID get all project heads by projectID
+func (s *store) GetByProjectID(projectID string) ([]*model.ProjectHead, error) {
+	var projectHeads []*model.ProjectHead
+	return projectHeads, s.db.Where("project_id = ?", projectID).Find(&projectHeads).Error
```
[the remark this answers is not in the document]
Comment (nguyennh4522) [reply pkg/store/project_head/project.go]: updated

File: pkg/store/project_slot/project_slot.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+// All get all projects with filter and pagination
+func (s *store) All(input GetListProjectSlotInput, pagination model.Pagination) ([]*model.ProjectSlot, int64, error) {
+	query := s.db.Table("project_slots").Where("project_slots.deleted_at IS NULL")
+	var total int64
+
+	query = query.Where("project_slots.project_id = ?", input.ProjectID).
+		Joins("LEFT JOIN project_members pm ON pm.project_slot_id = project_slots.id AND pm.project_id = ?", input.ProjectID)
+
+	switch input.Status {
+	case string(model.ProjectMemberStatusPending):
```
[the remark this answers is not in the document]
Comment (nguyennh4522) [reply pkg/store/project_slot/project_slot.go]: updated

File: pkg/store/project_slot/project_slot.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+		query = query.Where("pm.id IS NULL")
+
+	case string(model.ProjectMemberStatusInactive):
```
[the remark this answers is not in the document]
Comment (nguyennh4522) [reply pkg/store/project_slot/project_slot.go]: updated

File: pkg/store/project_slot/project_slot.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+		query = query.Where("pm.joined_date IS NOT NULL AND pm.left_date IS NOT NULL AND pm.left_date < now()")
+
+	case string(model.ProjectMemberStatusActive):
```
[the remark this answers is not in the document]
Comment (nguyennh4522) [reply pkg/store/project_slot/project_slot.go]: updated

File: pkg/view/project.go (hunk 4 above)
Comment (nguyennh4522) [reply pkg/view/project.go]: updated

File: pkg/store/project_slot/filter.go (hunk 3 above)
Comment (nguyennh4522) [reply pkg/store/project_slot/filter.go]: updated

File: pkg/store/project/project.go [hunk 11]
```diff
@@ -69,8 +69,7 @@ func (s *store) Create(db *gorm.DB, project *model.Project) error {
...
 }
 
-type GetListProjectInput struct {
-	Status string `json:"status"`
-	Name   string `json:"name"`
-	Type   string `json:"type"`
+// Exist return true/false if project exist or not
+func (s *store) Exist(db *gorm.DB, id string) bool {
+	return db.Where("id = ?", id).First(&model.Project{}).Error == nil
 }
```
Comment (namnhce) [inline pkg/store/project/project.go:75]:
```suggestion
func (s *store) Exist(db *gorm.DB, id string) (bool, error) {
	type res struct {
		Result bool
	}

	result := res{}
	query := db.DB().Raw("SELECT EXISTS (SELECT * FROM projects WHERE id = ?) as result", id)

	return result.Result, query.Scan(&result).Error
}
```

File: pkg/store/project/project.go [hunk 12]
```diff
@@ -69,8 +69,7 @@ func (s *store) Create(db *gorm.DB, project *model.Project) error {
 	return db.Create(&project).Error
 }
 
-type GetListProjectInput struct {
-	Status string `json:"status"`
-	Name   string `json:"name"`
-	Type   string `json:"type"`
+// Exist return true/false if project exist or not
+func (s *store) Exist(db *gorm.DB, id string) bool {
```
Comment (namnhce) [inline pkg/store/project/project.go:73]: If you check it by comparing `error` with `nil`, we possibly missing some errors caused by DBMS

**Review by namnhce — CHANGES_REQUESTED** [review]

File: pkg/handler/project/project.go [hunk 13]
```diff
@@ -226,3 +228,82 @@ func (h *handler) Create(c *gin.Context) {
...
+	exist, err := h.store.Project.Exist(h.repo.DB(), projectID)
+	if err != nil {
+		l.Error(err, "failed to check project existence")
+		c.JSON(http.StatusInternalServerError, view.CreateResponse[any](nil, nil, err, query))
+		return
+	}
+
+	if !exist {
+		l.Info("project not found")
+		c.JSON(http.StatusNotFound, view.CreateResponse[any](nil, nil, errors.New("project not found"), nil))
```
Comment (namnhce) [inline pkg/handler/project/project.go:286]: tạo 1 biến error mới em
errors.New("project not found")

**Review by namnhce — CHANGES_REQUESTED** [review]

File: pkg/store/project/interface.go [hunk 14]
```diff
@@ -10,4 +10,5 @@ type IStore interface {
 	All(db *gorm.DB, input GetListProjectInput, pagination model.Pagination) ([]*model.Project, int64, error)
 	UpdateStatus(db *gorm.DB, projectID string, projectStatus model.ProjectStatus) (*model.Project, error)
 	Create(db *gorm.DB, project *model.Project) error
+	Exist(db *gorm.DB, id string) (bool, error)
```
Comment (namnhce) [inline pkg/store/project/interface.go:13]:
```suggestion
	Exists(db *gorm.DB, id string) (bool, error)
```

**Review by namnhce — CHANGES_REQUESTED** [review]

File: pkg/handler/project/project.go (hunk 13 above)
Comment (nguyennh4522) [reply pkg/handler/project/project.go]: updated

File: pkg/store/project/project.go (hunk 11 above)
Comment (nguyennh4522) [reply pkg/store/project/project.go]: updated

**Pull request merged**